python/pyarrow/meson.build

```diff
@@ -0,0 +1,475 @@
+        arrow_dataset_dep = []
+    else
+        arrow_dataset_dep = dependency(
+            'arrow-dataset',
+            'ArrowDataset',
+            modules: [f'ArrowDataset::arrow_dataset_@cmake_suffix@'],
+        )
+    endif
+    libpyarrow_deps += [arrow_dataset_dep]
+    cython_modules += {'_dataset': {'dependencies': arrow_dataset_dep}}
+endif
+
+if needs_flight
+    if get_option('sdist')
+        arrow_flight_dep = []
+    else
+        arrow_flight_dep = dependency(
+            'arrow-flight',
+            'ArrowFlight',
+            modules: [f'ArrowFlight::arrow_flight_@cmake_suffix@'],
+        )
+    endif
+    libpyarrow_deps += [arrow_flight_dep]
+endif
+
+if get_option('gandiva').enabled()
+    if get_option('sdist')
+        gandiva_dep = []
+    else
+        gandiva_dep = dependency(
+            'gandiva',
+            'Gandiva',
+            modules: [f'Gandiva::gandiva_@cmake_suffix@'],
+        )
+    endif
+    libpyarrow_deps += [gandiva_dep]
+    cython_modules += {'gandiva': {'dependencies': gandiva_dep}}
+endif
+
+if get_option('gcs').enabled()
+    cython_modules += {'_gcsfs': {}}
+endif
+
+if get_option('hdfs').enabled()
+    cython_modules += {'_hdfs': {}}
+endif
+
+if get_option('orc').enabled()
+    cython_modules += {'_orc': {}}
+
+    if needs_dataset
+        cython_modules += {'_dataset_orc': {'dependencies': arrow_dataset_dep}}
+    endif
+endif
+
+if get_option('s3').enabled()
+    cython_modules += {'_s3fs': {}}
+endif
+
+if needs_substrait
+    if get_option('sdist')
+        arrow_substrait_dep = []
+    else
+        arrow_substrait_dep = dependency(
+            'arrow-substrait',
+            'ArrowSubstrait',
+            modules: [f'ArrowSubstrait::arrow_substrait_@cmake_suffix@'],
+        )
+    endif
+    libpyarrow_deps += [arrow_substrait_dep]
+    cython_modules += {'_substrait': {'dependencies': arrow_substrait_dep}}
+endif
+
+needs_parquet = get_option('parquet').enabled()
+needs_parquet_encryption = get_option('parquet_require_encryption').enabled()
+if needs_parquet_encryption and not needs_parquet
+    warning(
+        '''
+        Building PyArrow with Parquet Encryption is requested, but Parquet
+        itself is not enabled. Ignoring the Parquet Encryption setting.,
+    ''',
+    )
+    needs_parquet_encryption = false
+endif
+
+if needs_parquet
+    if get_option('sdist')
+        parquet_dep = []
+    else
+        parquet_dep = dependency(
+            'parquet',
+            'Parquet',
+            modules: [f'Parquet::parquet_@cmake_suffix@'],
+        )
+    endif
+    libpyarrow_deps += [parquet_dep]
+    cython_modules += {'_parquet': {'dependencies': parquet_dep}}
+
+    if needs_dataset
+        cython_modules += {
+            '_dataset_parquet': {
+                'dependencies': [parquet_dep, arrow_dataset_dep],
+            },
+        }
+    endif
+endif
+
+gnu_symbol_visibility = host_machine.system() == 'darwin' ? 'default' : 'inlineshidden'
```

Comment:
On my local box using `inlineshidden` reduces the overall size of shared libraries by 5%. On macOS it was segfaulting, which from cursory research has to do with how the PyMod_INIT function is exposed